lib/upsert.rb

```diff
@@ -37,6 +37,31 @@ def logger
       end
     end
 
+    def mutex_for_row(upsert, row)
+      retrieve_mutex(upsert.table_name, row.selector.keys)
+    end
+
+    def mutex_for_function(upsert, row)
+      retrieve_mutex(upsert.table_name, row.selector.keys, row.setter.keys)
+    end
+
+    # TODO: Rewrite this to use the thread_safe gem, perhaps?
+    def retrieve_mutex(*args)
```

Comment:
you know, there is an argument that we should just drop all this complicated mutex stuff and change how it's invoked to make it thread safe

Reply:
That's very true, maybe returning an upsert instance based on the current thread, or something?  That's the problem when using class methods ...